ldp/nn/handlers/transformer_handler.py

```diff
@@ -0,0 +1,1007 @@
+
+    def __del__(self) -> None:
+        dist.destroy_process_group()
+
+
+class ParallelAsyncTransformer(AsyncTransformerInterface):
+    def __init__(self, config: TransformerHandlerConfig):
+        self._initialized = False
+
+        parallel_mode_config = config.parallel_mode_config
+        if not parallel_mode_config:
+            raise ValueError("Parallel mode config must be provided.")
+        self.config = config
+        self.tokenizer = config.lm_config.get_tokenizer()
+        maybe_set_tokenizer_chat_template(self.tokenizer)
+        self.tokenizer.pad_token = self.tokenizer.eos_token
+
+        match parallel_mode_config.execution_mode:
+            case ExecutionMode.LOCAL_MACHINE:
+                self._init_local_cluster(config, parallel_mode_config)
+            case ExecutionMode.SLURM_CLUSTER:
+                self._init_slurm_cluster(config, parallel_mode_config)
+            case _:
+                raise NotImplementedError(
+                    f"Unsupported execution mode: {parallel_mode_config.execution_mode}"
+                )
+
+        self._initialized = True
+
+        # don't call AsyncTorchModule.__init__ because we don't need to set up module[_call_fn]
+        AsyncBufferedWorker.__init__(
+            self,
+            batch_size=config.batch_size,
+            max_wait_interval=config.max_wait_interval,
+            collate_fn=config.collate_fn,
+            decollate_fn=config.decollate_fn,
+        )
+
+        def handler_call_fn(handler: ParallelTransformerHandler, *args, **kwargs):
+            return config.module_call_fn(handler.module, *args, **kwargs)
+
+        self.handler_call_fn = handler_call_fn
+
+    def _init_local_cluster(self, config, parallel_mode_config):
+        """Initialize a Dask cluster on local machine."""
+        self.cluster = LocalCUDACluster(
+            n_workers=parallel_mode_config.num_workers,
+            threads_per_worker=parallel_mode_config.num_cpus_per_worker,
+            host=parallel_mode_config.scheduler_addr,
+            port=parallel_mode_config.scheduler_port,
+            memory_limit=None,  # do not let Dask manage memory - if we OOM, we OOM
+        )
+        self._initialize_workers(config, parallel_mode_config)
+
+    def _init_slurm_cluster(self, config, parallel_mode_config):
+        """Initialize a SLURM-based Dask cluster with GPU allocation."""
+        self.cluster = SLURMCluster(
+            cores=parallel_mode_config.num_cpus_per_worker,
+            memory=parallel_mode_config.memory,
+            processes=1,  # Single dask worker per slurm worker
+            walltime=parallel_mode_config.walltime,
+            job_extra=[
+                "--gres=gpu:1"
+            ],  # 1 GPU per worker seems to be the common case for now
+            log_directory=parallel_mode_config.log_directory,
+        )
+        self._initialize_workers(config, parallel_mode_config)
+
+    def _initialize_workers(self, config, parallel_mode_config):
```

Comment:
```suggestion
    def _initialize_workers(self, dask_client_config, parallel_mode_config):
```

Having this take a `config` and a `parallel_mode_config` makes one wonder, what is this first config? Let's be more verbose on the name for the first `config`

Reply:
It's not a `dask_client_config` though. I have added typing to avoid ambiguity.